### PR TITLE
[react-onclickoutside] enable/disableOnClickOutside returned to optional

### PR DESCRIPTION
**Pull request by nicolashahn**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 

https://github.com/Pomax/react-onclickoutside/issues/171 and https://github.com/DefinitelyTyped/DefinitelyTyped/commit/0d96c339bccb58d2f37181411438ddb403c08f70

__tl;dr:__ changing the types of these methods from optional to required caused breakage for me and many other people, this merely reverts that change.

*(I don't think the the following two are applicable for this PR - please correct me if I'm wrong)*
- [ ] Increase the version number in the header if appropriate. 
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

I've tested by editing `node_modules/@types/react-onclickoutside/index.d.ts` in a local app as well to get this back to expected functionality, and it works.

### CODE REVIEW

File: types/react-onclickoutside/index.d.ts [hunk 1]
```diff
@@ -11,8 +11,8 @@ export interface HandleClickOutside<T> {
 }
 
 export interface InjectedOnClickOutProps {
-    disableOnClickOutside(): void;
-    enableOnClickOutside(): void;
+    disableOnClickOutside?(): void;
+    enableOnClickOutside?(): void;
```
Comment (LKay) [inline types/react-onclickoutside/index.d.ts]: These properties are not optional but are injected by component inside HoC and will always be present: https://github.com/Pomax/react-onclickoutside/blob/master/src/index.js#L201-L202

**Review by LKay — CHANGES_REQUESTED** [review]

File: types/react-onclickoutside/index.d.ts (hunk 1 above)
Comment (nicolashahn) [reply types/react-onclickoutside/index.d.ts]: Can you explain why I and several others are getting the error shown here, then? https://github.com/Pomax/react-onclickoutside/issues/171

If I don't have my component extend `InjectedOnClickOutProps` then I get the error in the first post on that link, but if I do I get

```
ERROR in [at-loader] ./source/components/search-react/index.tsx:184:35 
    TS2322: Type '{ allBrands: Brand[]; selectedBrands: () => Brand[]; isSelected: (brand: Brand) => boolean; hideB...' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<BrandsDropdownProps & OnClickOutProps, C...'.
  Type '{ allBrands: Brand[]; selectedBrands: () => Brand[]; isSelected: (brand: Brand) => boolean; hideB...' is not assignable to type 'Readonly<BrandsDropdownProps & OnClickOutProps>'.
    Property 'disableOnClickOutside' is missing in type '{ allBrands: Brand[]; selectedBrands: () => Brand[]; isSelected: (brand: Brand) => boolean; hideB...'.
```

So it doesn't appear to be automatically being injected.

Comment (LKay) [reply types/react-onclickoutside/index.d.ts]: It's probably the way how oyu define your inner component in Typescript. Seems like your props type don't match and they are expected to extend `InjectedOnClickProps`. This is the example of how it should loook more or less (it's a little verbose to get the whole picture):

```typescript
import * as React form "react";
import onClickOutside form "react-onclickoutside";

type InjectedProps = InjectedOnClickProps;

interface OwnProps {
  foo?: string;
}

type MyComponentProps = OwnProps & InjectedProps;

// stateless component case
const MyComponentStateless = onClickOutside<OwnProps>(({  
    foo,
    enableOnClickOutside,
    disableOnClickOutside
}) => {
   disableClickOutside();
   return <div>{ foo }</div>;
});

// class component case
const MyComponentClass = onClickOutside(
    class extends React.Component<MyComponentProps> {
        componentWillMount() {
            this.props.disableClickOutside();
        }

        render() {
            return <div>{ this.props.foo }</div>;
        }
    }
)
```

In that case the types match and you can access injected props inside your component, yet outer component only accepts outer props which will equals to `OwnProps`.

Comment (nicolashahn) [reply types/react-onclickoutside/index.d.ts]: Thank you, I was expecting this to work: 
```
interface OwnProps extends InjectedOnClickProps {
    ...
}

class MyComponent extends React.Component<OwnProps, {}> { 
    ...
    constructor (props: OwnProps) {
    ...
```

Switching to using the `&` operator had it working.

```
interface OwnProps {...}

class MyComponent extends React.Component<OwnProps & InjectedOnClickProps, {}> {
    ...
    constructor (props: OwnProps & InjectedOnClickProps) {
    ...
```

Will close this.

Comment (davishmcclurg) [reply types/react-onclickoutside/index.d.ts]: @LKay do you have any insight on why extending `InjectedOnClickProps` doesn't work as described above? I'm still confused and haven't been able to figure out why an intersection type behaves differently from the extended interface.
